### PR TITLE
Add Solana minimum balance library

### PR DESCRIPTION
**Pull request by seanyoung**
Add runtime function for calculating the balance required for instantiating a contract of a certain size.

### CODE REVIEW

File: docs/targets/solana.rst [hunk 1]
```diff
@@ -452,6 +452,14 @@ and `system_instruction.spec.ts <https://github.com/hyperledger/solang/blob/main
 
 .. _account_management:
 
+Minimum balance
++++++++++++++++
```
Comment (LucasSte) [inline docs/targets/solana.rst:456]: This section is misplaced. `.. _account_management:` should be strictly above `Solana Account Management` section.

File: docs/targets/solana.rst [hunk 2]
```diff
@@ -452,6 +452,14 @@ and `system_instruction.spec.ts <https://github.com/hyperledger/solang/blob/main
...
+
+In order to instantiate a contract, you need the minimum balance required for an Solana account for a given
+size.  There is a function `minimum_balance(uint64 space)` defined in
```
Comment (LucasSte) [inline docs/targets/solana.rst:459]:
```suggestion
In order to instantiate a contract, you need the minimum balance required for a Solana account of a given
size.  There is a function `minimum_balance(uint64 space)` defined in
```

File: docs/targets/solana.rst [hunk 3]
```diff
@@ -452,6 +452,14 @@ and `system_instruction.spec.ts <https://github.com/hyperledger/solang/blob/main
...
+`minimum_balance.sol <https://github.com/hyperledger/solang/blob/main/solana-library/minimum_balance.sol>`_
+which calculates this.
```
Comment (LucasSte) [inline docs/targets/solana.rst:461]:
```suggestion
that calculates this.
```

File: solana-library/minimum_balance.sol [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+/// This calculation is based on:
+/// - 10^9 lamports per SOL
+/// - $1 per SOL
+/// - $0.01 per megabyte day
+/// - $3.65 per megabyte year
+uint64 constant DEFAULT_LAMPORTS_PER_BYTE_YEAR = 1_000_000_000 / 100 * 365 / (1024 * 1024);
+
+/// Default amount of time (in years) the balance has to include rent for the
+/// account to be rent exempt.
+/// Note that rent is currently not
```
Comment (LucasSte) [inline solana-library/minimum_balance.sol:17]: `is currently not` what?

File: solana-library/minimum_balance.sol [hunk 5]
```diff
@@ -0,0 +1,30 @@
+// SPDX-License-Identifier: Apache-2.0
+
+// Disclaimer: This library provides functions for working with storage rent. Although it is production ready,
+// it has not been audited for security, so use it at your own risk.
+
+/// Default rental rate in lamports/byte-year.
+///
+/// This calculation is based on:
+/// - 10^9 lamports per SOL
+/// - $1 per SOL
```
Comment (xermicus) [inline solana-library/minimum_balance.sol:10]: I don't know too much about Solana. But $1 per SOL seems widely off. Where is this number coming from? Would be good to provide the source

File: docs/targets/solana.rst (hunk 1 above)
Comment (seanyoung) [reply docs/targets/solana.rst:456]: Good point, I don't know how I missed that. Fixed

File: solana-library/minimum_balance.sol (hunk 4 above)
Comment (seanyoung) [reply solana-library/minimum_balance.sol:17]: I don't know how that crept in. Thanks for spotting.

File: solana-library/minimum_balance.sol (hunk 5 above)
Comment (xermicus) [reply solana-library/minimum_balance.sol:14]: I see, thanks

**Review by xermicus — APPROVED** [review]

File: docs/targets/solana.rst [hunk 6]
```diff
@@ -450,6 +450,14 @@ contracts on chain. Examples are available on Solang's integration tests.
 See `system_instruction_example.sol <https://github.com/hyperledger/solang/blob/main/integration/solana/system_instruction_example.sol>`_
 and `system_instruction.spec.ts <https://github.com/hyperledger/solang/blob/main/integration/solana/system_instruction.spec.ts>`_
 
+Minimum balance
++++++++++++++++
+
+In order to instantiate a contract, you need the minimum balance required for a Solana account of a given
+size.  There is a function ``minimum_balance(uint64 space)`` defined in
+`minimum_balance.sol <https://github.com/hyperledger/solang/blob/main/solana-library/minimum_balance.sol>`_
+that calculates this.
```
Comment (LucasSte) [inline docs/targets/solana.rst:459]: I've re-read this sentence and, although it is correctly written, I thought that `to calculate this` sounds better. This is more style, so only change if you share my opinion.

File: integration/solana/minbalance.sol [hunk 7]
```diff
@@ -0,0 +1,16 @@
+import '../../solana-library/minimum_balance.sol';
+
+contract minbalance {
+    /// Anchor requires view functions to have a return value
+    function test1() pure public returns (bool) {
```
Comment (LucasSte) [inline integration/solana/minbalance.sol:5]: Isn't this a pure function?

File: solana-library/minimum_balance.sol [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+/// - $0.01 per megabyte day
+/// - $3.65 per megabyte year
+uint64 constant DEFAULT_LAMPORTS_PER_BYTE_YEAR = 1_000_000_000 / 100 * 365 / (1024 * 1024);
+
+/// Default amount of time (in years) the balance has to include rent for the
+/// account to be rent exempt.
```
Comment (LucasSte) [inline solana-library/minimum_balance.sol:20]: This description is a bit unclear.

File: solana-library/minimum_balance.sol [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+uint64 constant DEFAULT_EXEMPTION_THRESHOLD = 2;
+
+/// Account storage overhead for calculation of base rent.
+///
+/// This is the number of bytes required to store an account with no data. It is
+/// added to an accounts data length when calculating [`Rent::minimum_balance`].
```
Comment (LucasSte) [inline solana-library/minimum_balance.sol:26]:
```suggestion
/// added to an account's data length when calculating [`Rent::minimum_balance`].
```

File: docs/targets/solana.rst (hunk 6 above)
Comment (seanyoung) [reply docs/targets/solana.rst:459]: I agree, that sounds better

File: integration/solana/minbalance.sol (hunk 7 above)
Comment (seanyoung) [reply integration/solana/minbalance.sol:5]: Anchor does not distinguish between `pure` and `view`.

Comment (seanyoung) [reply integration/solana/minbalance.sol:5]: I've improved the comment

File: solana-library/minimum_balance.sol (hunk 8 above)
Comment (seanyoung) [reply solana-library/minimum_balance.sol:20]: I've reworded it. Note this is all academy since rent is not implemented yet.

**Pull request merged**